textutils-2.1/man/cut.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.28.
.TH CUT "1" "July 2002" "cut (textutils) 2.1" "User Commands"
.SH NAME
cut \- remove sections from each line of files
.SH SYNOPSIS
.B cut
[\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Print selected parts of lines from each FILE to standard output.
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-b\fR, \fB\-\-bytes\fR=\fILIST\fR
output only these bytes
.TP
\fB\-c\fR, \fB\-\-characters\fR=\fILIST\fR
output only these characters
.TP
\fB\-d\fR, \fB\-\-delimiter\fR=\fIDELIM\fR
use DELIM instead of TAB for field delimiter
.TP
\fB\-f\fR, \fB\-\-fields\fR=\fILIST\fR
output only these fields;  also print any line
that contains no delimiter character, unless
the \fB\-s\fR option is specified
.TP
\fB\-n\fR
(ignored)
.TP
\fB\-s\fR, \fB\-\-only\-delimited\fR
do not print lines not containing delimiters
.TP
\fB\-\-output\-delimiter\fR=\fISTRING\fR
use STRING as the output delimiter
the default is to use the input delimiter
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
Use one, and only one of \fB\-b\fR, \fB\-c\fR or \fB\-f\fR.  Each LIST is made up of one
range, or many ranges separated by commas.  Each range is one of:
.TP
N
N'th byte, character or field, counted from 1
.TP
N-
from N'th byte, character or field, to end of line
.TP
N-M
from N'th to M'th (included) byte, character or field
.TP
\fB\-M\fR
from first to M'th (included) byte, character or field
.PP
With no FILE, or when FILE is -, read standard input.
.SH AUTHOR
Written by David Ihnat, David MacKenzie, and Jim Meyering.
.SH "REPORTING BUGS"
Report bugs to <bug-textutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2002 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B cut
is maintained as a Texinfo manual.  If the
.B info
and
.B cut
programs are properly installed at your site, the command
.IP
.B info cut
.PP
should give you access to the complete manual.


zsh/ChangeLog
beta 0.75
Wed Jun 14 14:47:46 2000    <amold@SKYNYRD>
	* ntport\dirent.c: Fix serious bug in wildcard expansion on Win9x 
beta 0.74
Mon Jan 31 14:53:06 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed bug in converting '\'s in $HOME 
	to '/'s.
beta 0.73
Mon Jan 17 11:55:15 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed crash in long arglist introduced by
 	0.70
beta 0.72
Tue Jan 11 08:28:29 2000    <amold@SKYNYRD>
	* ntport\support.c: Fixed memory corruption on Win9x
beta 0.71
Mon Jan 10 11:49:39 2000    <amold@SKYNYRD>
	* ntport\stdio.c: Fix bug in inheritance of file handles.  
beta 0.69,0.70
Fri Dec 17 11:53:19 1999    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in argv[0] of spawned argument. Might
	have manifested as applications in ZSHROOT failing to spawn 
	themselves. 
	Also added winntnoquoteprotect option. Use with tools like MKS
	if you see unexpected quoting of "
beta 0.68
Mon Sep 27 11:20:24 1999    <amold@SKYNYRD>
	* ntport\support.c: support paths > 1024 characters. 
beta 0.67
Tue Sep 07 10:05:02 1999    <amold@SKYNYRD>
	* ntport\io.c: Fix bug that affected FPATH/autoload 
beta 0.65
Fri Aug 27 11:41:58 1999    <amold@SKYNYRD>
	* ntport\dirent.c: Fix completion with drive letters in string. 
beta 0.64
Fri Aug 27 11:04:11 1999    <amold@SKYNYRD>
	* ntport\io.c: Unfix Win95 spurious linefeed fix for Windows 98. 
beta 0.63
Wed May 19 16:21:37 1999    <amold@SKYNYRD>
	* exec.c, misc: VC 6.0 compliation errors. Disable optmizations
	in functions that call fork() to avoid losing global variables
	that might be in registers.
beta 0.62
Tue Apr 13 17:06:12 1999    <amold@SKYNYRD>
	* ntport\support.c, zle_misc.c: Keep server name in prompt
 	if %C is specified and pwd is UNC.
beta 0.61
Tue Mar 23 12:17:50 1999  unknown  <bogus@SKYNYRD>
	* ntport\support.c: bug fix for previous fix. 
beta 0.60
Sat Feb 13 17:49:14 1999    <amold@SKYNYRD>
	* ntport\support.c,ntbin.c: Fix lame arg-concatenation in
 	exec and start paths. commands with large argument strings should
	now execute *much* faster.
beta 0.57
Tue Feb 02 18:09:39 1999    <amold@SKYNYRD>
	* ntport\ntbin.c: fixed bug that improperly passed arguments
	to applications launched by windows "shell" associations. 
	For example, .bat files.
beta 0.55
Mon Oct 26 11:09:44 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: start builtin will not use '/' for option
 	flags. only '-' will work.
beta 0.54
Fri Oct 16 12:52:08 1998    <amold@SKYNYRD>
	* ntport\support.c: added WINNTWAITFORGUIAPPS option. 
beta 0.53
Mon Oct 12 12:31:59 1998    <amold@SKYNYRD>
	* init.c: look for "zshrc" as well as ".zshrc" 
beta 0.52
Mon Sep 28 17:31:08 1998    <amold@SKYNYRD>
	* utils.c: attempt to ignore case while correcting spelling
 	if WINNTIGNORECASE is set.
beta 0.51
Thu Sep 17 13:48:27 1998    <amold@SKYNYRD>
	* hashtable.c,support.c: filenames without extensions and
 	*only* files with extensions specified in the PATHEXT variable
 	will be hashed. The default extensions are ".com;.exe;.cmd;.bat"
beta 0.50
Wed Sep 16 15:58:56 1998    <amold@SKYNYRD>
	* hashtable.c: added hashing for external exes without
	extension. bogus spelling corrections should be reduced.
beta 0.49
Tue Sep 15 11:09:31 1998    <amold@SKYNYRD>
	* support.c: The shell will automatically set TERM to vt100
 	if it's not set in the environment. There should be no need
	to set the variable explicitly any more.
beta 0.48
Mon Aug 03 17:47:08 1998    <amold@SKYNYRD>
	* support.c: Fixed bug in converting paths to '/'-delimited 
beta 0.47
Mon Aug 03 16:59:38 1998    <amold@SKYNYRD>
	* utils.c: Fixed bug in applying previous patch. If %~ wasn't
	working, this should fix it.
beta 0.46
Sat Jul 18 14:36:10 1998    <amold@SKYNYRD>
	* utils.c,globals.h,zsh.h: patch from dominik for 
	winntconvertbackslash.
beta 0.45
Mon Jun 29 11:30:55 1998    <amold@SKYNYRD>
	* utils.c: patch from dominik vogt for honoring winntignorecase
 	in %~ prompt.
beta 0.44
Mon Jun 22 11:19:12 1998    <amold@SKYNYRD>
	* support.c,io.c: moved some startup code around. fixed spurious
 	linefeeds on win95
beta 0.43
Tue Jun 09 17:25:29 1998    <amold@SKYNYRD>
	* builtin.c,ntport\io.c: Fix for the read builtin not
 	echoing characters
beta 0.429
Thu Jun 04 17:43:54 1998    <amold@SKYNYRD>
	* utils.c: Fixed %~ prompt spec.
beta 0.428
Thu Jun 04 16:50:21 1998    <amold@SKYNYRD>
	* builtin.c: Fixed problems with the setting of PWD.
beta 0.427
Thu May 28 14:31:28 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to let apps with spaces in their path
	run. c:\program files\myapp.exe , for example.
beta 0.426
Wed May 27 10:46:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed stupid command buffer overflow
 	problem. If you crashed while grepping through a bunch of
	files, this should fix it.
beta 0.425
Fri May 22 08:55:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in parsing args to child processes.
beta 0.424
Sun May 17 13:18:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that caused zsh to spawn a new
	console if $path contained a component '/'. 
beta 0.423
Fri May 15 16:08:17 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: winntlamepathfix now works with the 'start'
	builtin.
beta 0.421
Mon Apr 06 16:18:20 1998    <amold@SKYNYRD>
	* ntport\support.c,fork.c: Fixed bug in heap initialization for
	forked children that crashed the child process.
beta 0.42
Sun Apr 05 13:36:49 1998    <amold@SKYNYRD>
	* ntport\support.c: Put in code to fix the case of a directory
 	you cd into. This is extremely expensive and really silly, so
	it may become optional in the future.
beta 0.41
Sat Apr 04 15:27:10 1998    <amold@SKYNYRD>

textutils-2.1/src/cut.c
/* cut - remove parts of lines of files
   Copyright (C) 1984, 1997, 1998, 1999, 2000, 2001 by David M. Ihnat

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Written by David Ihnat.  */

/* POSIX changes, bug fixes, long-named options, and cleanup
   by David MacKenzie <djm@gnu.ai.mit.edu>.

   Rewrite cut_fields and cut_bytes -- Jim Meyering.  */

#include <config.h>

#include <stdio.h>
#include <assert.h>
#include <getopt.h>
#include <sys/types.h>
#include "system.h"
#include "getstr.h"
#include "closeout.h"
#include "error.h"

/* The official name of this program (e.g., no `g' prefix).  */
#define PROGRAM_NAME "cut"

#define AUTHORS N_ ("David Ihnat, David MacKenzie, and Jim Meyering")

#define FATAL_ERROR(Message)						\
  do									\
    {									\
      error (0, 0, (Message));						\
      usage (2);							\
    }									\
  while (0)

/* Append LOW, HIGH to the list RP of range pairs, allocating additional
   space if necessary.  Update local variable N_RP.  When allocating,
   update global variable N_RP_ALLOCATED.  */

#define ADD_RANGE_PAIR(rp, low, high)					\
  do									\
    {									\
      if (n_rp >= n_rp_allocated)					\
	{								\
	  n_rp_allocated *= 2;						\
	  (rp) = (struct range_pair *) xrealloc ((char *) (rp),		\
				   n_rp_allocated * sizeof (*(rp)));	\
	}								\
      rp[n_rp].lo = (low);						\
      rp[n_rp].hi = (high);						\
      ++n_rp;								\
    }									\
  while (0)

struct range_pair
  {
    unsigned int lo;
    unsigned int hi;
  };

/* This buffer is used to support the semantics of the -s option
   (or lack of same) when the specified field list includes (does
   not include) the first field.  In both of those cases, the entire
   first field must be read into this buffer to determine whether it
   is followed by a delimiter or a newline before any of it may be
   output.  Otherwise, cut_fields can do the job without using this
   buffer.  */
static char *field_1_buffer;

/* The number of bytes allocated for FIELD_1_BUFFER.  */
static size_t field_1_bufsize;

/* The largest field or byte index used as an endpoint of a closed
   or degenerate range specification;  this doesn't include the starting
   index of right-open-ended ranges.  For example, with either range spec
   `2-5,9-', `2-3,5,9-' this variable would be set to 5.  */
static unsigned int max_range_endpoint;

/* If nonzero, this is the index of the first field in a range that goes
   to end of line. */
static unsigned int eol_range_start;

/* In byte mode, which bytes to output.
   In field mode, which DELIM-separated fields to output.
   Both bytes and fields are numbered starting with 1,
   so the zeroth element of this array is unused.
   A field or byte K has been selected if
   (K <= MAX_RANGE_ENDPOINT and PRINTABLE_FIELD[K])
    || (EOL_RANGE_START > 0 && K >= EOL_RANGE_START).  */
static int *printable_field;

enum operating_mode
  {
    undefined_mode,

    /* Output characters that are in the given bytes. */
    byte_mode,

    /* Output the given delimeter-separated fields. */
    field_mode
  };

/* The name this program was run with. */
char *program_name;

static enum operating_mode operating_mode;

/* If nonzero do not output lines containing no delimeter characters.
   Otherwise, all such lines are printed.  This option is valid only
   with field mode.  */
static int suppress_non_delimited;

/* The delimeter character for field mode. */
static int delim;

/* The length of output_delimiter_string.  */
static size_t output_delimiter_length;

/* The output field separator string.  Defaults to the 1-character
   string consisting of the input delimiter.  */
static char *output_delimiter_string;

/* Nonzero if we have ever read standard input. */
static int have_read_stdin;

/* For long options that have no equivalent short option, use a
   non-character as a pseudo short option, starting with CHAR_MAX + 1.  */
enum
{
  OUTPUT_DELIMITER_OPTION = CHAR_MAX + 1
};

static struct option const longopts[] =
{
  {"bytes", required_argument, 0, 'b'},
  {"characters", required_argument, 0, 'c'},
  {"fields", required_argument, 0, 'f'},
  {"delimiter", required_argument, 0, 'd'},
  {"only-delimited", no_argument, 0, 's'},
  {"output-delimiter", required_argument, 0, OUTPUT_DELIMITER_OPTION},
  {GETOPT_HELP_OPTION_DECL},
  {GETOPT_VERSION_OPTION_DECL},
  {0, 0, 0, 0}
};

void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION]... [FILE]...\n\
"),
	      program_name);
      fputs (_("\
Print selected parts of lines from each FILE to standard output.\n\
\n\
"), stdout);
      fputs (_("\
Mandatory arguments to long options are mandatory for short options too.\n\
"), stdout);
      fputs (_("\
  -b, --bytes=LIST        output only these bytes\n\
  -c, --characters=LIST   output only these characters\n\
  -d, --delimiter=DELIM   use DELIM instead of TAB for field delimiter\n\
"), stdout);
      fputs (_("\
  -f, --fields=LIST       output only these fields;  also print any line\n\
                            that contains no delimiter character, unless\n\
                            the -s option is specified\n\
  -n                      (ignored)\n\
"), stdout);
      fputs (_("\
  -s, --only-delimited    do not print lines not containing delimiters\n\
      --output-delimiter=STRING  use STRING as the output delimiter\n\
                            the default is to use the input delimiter\n\
"), stdout);
      fputs (HELP_OPTION_DESCRIPTION, stdout);
      fputs (VERSION_OPTION_DESCRIPTION, stdout);
      fputs (_("\
\n\
Use one, and only one of -b, -c or -f.  Each LIST is made up of one\n\
range, or many ranges separated by commas.  Each range is one of:\n\
\n\
  N     N'th byte, character or field, counted from 1\n\
  N-    from N'th byte, character or field, to end of line\n\
  N-M   from N'th to M'th (included) byte, character or field\n\
  -M    from first to M'th (included) byte, character or field\n\
\n\
With no FILE, or when FILE is -, read standard input.\n\
"), stdout);
      printf (_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
    }
  exit (status == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}

static int
print_kth (unsigned int k)
{
  return ((0 < eol_range_start && eol_range_start <= k)
	  || (k <= max_range_endpoint && printable_field[k]));
}

/* Given the list of field or byte range specifications FIELDSTR, set
   MAX_RANGE_ENDPOINT and allocate and initialize the PRINTABLE_FIELD
   array.  If there is a right-open-ended range, set EOL_RANGE_START
   to its starting index.  FIELDSTR should be composed of one or more
   numbers or ranges of numbers, separated by blanks or commas.
   Incomplete ranges may be given: `-m' means `1-m'; `n-' means `n'
   through end of line.  Return nonzero if FIELDSTR contains at least
   one field specification, zero otherwise.  */

/* FIXME-someday:  What if the user wants to cut out the 1,000,000-th field
   of some huge input file?  This function shouldn't have to alloate a table
   of a million ints just so we can test every field < 10^6 with an array
   dereference.  Instead, consider using a dynamic hash table.  It would be
   simpler and nearly as good a solution to use a 32K x 4-byte table with
   one bit per field index instead of a whole `int' per index.  */

static int
set_fields (const char *fieldstr)
{
  unsigned int initial = 1;	/* Value of first number in a range.  */
  unsigned int value = 0;	/* If nonzero, a number being accumulated.  */
  int dash_found = 0;		/* Nonzero if a '-' is found in this field.  */
  int field_found = 0;		/* Non-zero if at least one field spec
				   has been processed.  */

  struct range_pair *rp;
  unsigned int n_rp;
  unsigned int n_rp_allocated;
  unsigned int i;

  n_rp = 0;
  n_rp_allocated = 16;
  rp = (struct range_pair *) xmalloc (n_rp_allocated * sizeof (*rp));

  /* Collect and store in RP the range end points.
     It also sets EOL_RANGE_START if appropriate.  */

  for (;;)
    {
      if (*fieldstr == '-')
	{
	  /* Starting a range. */
	  if (dash_found)
	    FATAL_ERROR (_("invalid byte or field list"));
	  dash_found++;
	  fieldstr++;

	  if (value)
	    {
	      initial = value;
	      value = 0;
	    }
	  else
	    initial = 1;
	}
      else if (*fieldstr == ',' || ISBLANK (*fieldstr) || *fieldstr == '\0')
	{
	  /* Ending the string, or this field/byte sublist. */
	  if (dash_found)
	    {
	      dash_found = 0;

	      /* A range.  Possibilites: -n, m-n, n-.
		 In any case, `initial' contains the start of the range. */
	      if (value == 0)
		{
		  /* `n-'.  From `initial' to end of line. */
		  eol_range_start = initial;
		  field_found = 1;
		}
	      else
		{
		  /* `m-n' or `-n' (1-n). */
		  if (value < initial)
		    FATAL_ERROR (_("invalid byte or field list"));

		  /* Is there already a range going to end of line? */
		  if (eol_range_start != 0)
		    {
		      /* Yes.  Is the new sequence already contained
			 in the old one?  If so, no processing is
			 necessary. */
		      if (initial < eol_range_start)
			{
			  /* No, the new sequence starts before the
			     old.  Does the old range going to end of line
			     extend into the new range?  */
			  if (value + 1 >= eol_range_start)
			    {
			      /* Yes.  Simply move the end of line marker. */
			      eol_range_start = initial;
			    }
			  else
			    {
			      /* No.  A simple range, before and disjoint from
				 the range going to end of line.  Fill it. */
			      ADD_RANGE_PAIR (rp, initial, value);
			    }

			  /* In any case, some fields were selected. */
			  field_found = 1;
			}
		    }
		  else
		    {
		      /* There is no range going to end of line. */
		      ADD_RANGE_PAIR (rp, initial, value);
		      field_found = 1;
		    }
		  value = 0;
		}
	    }
	  else if (value != 0)
	    {
	      /* A simple field number, not a range. */
	      ADD_RANGE_PAIR (rp, value, value);
	      value = 0;
	      field_found = 1;
	    }

	  if (*fieldstr == '\0')
	    {
	      break;
	    }

	  fieldstr++;
	}
      else if (ISDIGIT (*fieldstr))
	{
	  /* FIXME: detect overflow?  */
	  value = 10 * value + *fieldstr - '0';
	  fieldstr++;
	}
      else
	FATAL_ERROR (_("invalid byte or field list"));
    }

  max_range_endpoint = 0;
  for (i = 0; i < n_rp; i++)
    {
      if (rp[i].hi > max_range_endpoint)
	max_range_endpoint = rp[i].hi;
    }

  /* Allocate an array large enough so that it may be indexed by
     the field numbers corresponding to all finite ranges
     (i.e. `2-6' or `-4', but not `5-') in FIELDSTR.  */

  printable_field = (int *) xmalloc ((max_range_endpoint + 1) * sizeof (int));
  memset (printable_field, 0, (max_range_endpoint + 1) * sizeof (int));

  /* Set the array entries corresponding to integers in the ranges of RP.  */
  for (i = 0; i < n_rp; i++)
    {
      unsigned int j;
      for (j = rp[i].lo; j <= rp[i].hi; j++)
	{
	  printable_field[j] = 1;
	}
    }

  free (rp);

  return field_found;
}

/* Read from stream STREAM, printing to standard output any selected bytes.  */

static void
cut_bytes (FILE *stream)
{
  unsigned int byte_idx;	/* Number of chars in the line so far. */

  byte_idx = 0;
  while (1)
    {
      register int c;		/* Each character from the file. */

      c = getc (stream);

      if (c == '\n')
	{
	  putchar ('\n');
	  byte_idx = 0;
	}
      else if (c == EOF)
	{
	  if (byte_idx > 0)
	    putchar ('\n');
	  break;
	}
      else
	{
	  ++byte_idx;
	  if (print_kth (byte_idx))
	    {
	      putchar (c);
	    }
	}
    }
}

/* Read from stream STREAM, printing to standard output any selected fields.  */

static void
cut_fields (FILE *stream)
{
  int c;
  unsigned int field_idx;
  int found_any_selected_field;
  int buffer_first_field;
  int empty_input;

  found_any_selected_field = 0;
  field_idx = 1;

  c = getc (stream);
  empty_input = (c == EOF);
  if (c != EOF)
    ungetc (c, stream);

  /* To support the semantics of the -s flag, we may have to buffer
     all of the first field to determine whether it is `delimited.'
     But that is unnecessary if all non-delimited lines must be printed
     and the first field has been selected, or if non-delimited lines
     must be suppressed and the first field has *not* been selected.
     That is because a non-delimited line has exactly one field.  */
  buffer_first_field = (suppress_non_delimited ^ !print_kth (1));

  while (1)
    {
      if (field_idx == 1 && buffer_first_field)
	{
	  int len;

	  len = getstr (&field_1_buffer, &field_1_bufsize, stream,
			delim, '\n', 0);
	  if (len < 0)
	    {
	      if (ferror (stream) || feof (stream))
		break;
	      xalloc_die ();
	    }

	  assert (len != 0);

	  /* If the first field extends to the end of line (it is not
	     delimited) and we are printing all non-delimited lines,
	     print this one.  */
	  if ((unsigned char) field_1_buffer[len - 1] != delim)
	    {
	      if (suppress_non_delimited)
		{
		  /* Empty.  */
		}
	      else
		{
		  fwrite (field_1_buffer, sizeof (char), len, stdout);
		  /* Make sure the output line is newline terminated.  */
		  if (field_1_buffer[len - 1] != '\n')
		    putchar ('\n');
		}
	      continue;
	    }
	  if (print_kth (1))
	    {
	      /* Print the field, but not the trailing delimiter.  */
	      fwrite (field_1_buffer, sizeof (char), len - 1, stdout);
	      found_any_selected_field = 1;
	    }
	  ++field_idx;
	}

      if (c != EOF)
	{
	  if (print_kth (field_idx))
	    {
	      if (found_any_selected_field)
		{
		  fwrite (output_delimiter_string, sizeof (char),
			  output_delimiter_length, stdout);
		}
	      found_any_selected_field = 1;

	      while ((c = getc (stream)) != delim && c != '\n' && c != EOF)
		{
		  putchar (c);
		}
	    }
	  else
	    {
	      while ((c = getc (stream)) != delim && c != '\n' && c != EOF)
		{
		  /* Empty.  */
		}
	    }
	}

      if (c == '\n')
	{
	  c = getc (stream);
	  if (c != EOF)
	    {
	      ungetc (c, stream);
	      c = '\n';
	    }
	}

      if (c == delim)
	++field_idx;
      else if (c == '\n' || c == EOF)
	{
	  if (found_any_selected_field
	      || (!empty_input && !(suppress_non_delimited && field_idx == 1)))
	    putchar ('\n');
	  if (c == EOF)
	    break;
	  field_idx = 1;
	  found_any_selected_field = 0;
	}
    }
}

static void
cut_stream (FILE *stream)
{
  if (operating_mode == byte_mode)
    cut_bytes (stream);
  else
    cut_fields (stream);
}

/* Process file FILE to standard output.
   Return 0 if successful, 1 if not. */

static int
cut_file (char *file)
{
  FILE *stream;

  if (STREQ (file, "-"))
    {
      have_read_stdin = 1;
      stream = stdin;
    }
  else
    {
      stream = fopen (file, "rb");
      if (stream == NULL)
	{
	  error (0, errno, "%s", file);
	  return 1;
	}
    }

  cut_stream (stream);

  if (ferror (stream))
    {
      error (0, errno, "%s", file);
      return 1;
    }
  if (STREQ (file, "-"))
    clearerr (stream);		/* Also clear EOF. */
  else if (fclose (stream) == EOF)
    {
      error (0, errno, "%s", file);
      return 1;
    }
  return 0;
}

int
main (int argc, char **argv)
{
  int optc, exit_status = 0;
  int delim_specified = 0;

  program_name = argv[0];
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  atexit (close_stdout);

  operating_mode = undefined_mode;

  /* By default, all non-delimited lines are printed.  */
  suppress_non_delimited = 0;

  delim = '\0';
  have_read_stdin = 0;

  while ((optc = getopt_long (argc, argv, "b:c:d:f:ns", longopts, NULL)) != -1)
    {
      switch (optc)
	{
	case 0:
	  break;

	case 'b':
	case 'c':
	  /* Build the byte list. */
	  if (operating_mode != undefined_mode)
	    FATAL_ERROR (_("only one type of list may be specified"));
	  operating_mode = byte_mode;
	  if (set_fields (optarg) == 0)
	    FATAL_ERROR (_("missing list of positions"));
	  break;

	case 'f':
	  /* Build the field list. */
	  if (operating_mode != undefined_mode)
	    FATAL_ERROR (_("only one type of list may be specified"));
	  operating_mode = field_mode;
	  if (set_fields (optarg) == 0)
	    FATAL_ERROR (_("missing list of fields"));
	  break;

	case 'd':
	  /* New delimiter. */
	  /* Interpret -d '' to mean `use the NUL byte as the delimiter.'  */
	  if (optarg[0] != '\0' && optarg[1] != '\0')
	    FATAL_ERROR (_("the delimiter must be a single character"));
	  delim = (unsigned char) optarg[0];
	  delim_specified = 1;
	  break;

	case OUTPUT_DELIMITER_OPTION:
	  /* Interpret --output-delimiter='' to mean
	     `use the NUL byte as the delimiter.'  */
	  output_delimiter_length = (optarg[0] == '\0'
				     ? 1 : strlen (optarg));
	  output_delimiter_string = xstrdup (optarg);
	  break;

	case 'n':
	  break;

	case 's':
	  suppress_non_delimited = 1;
	  break;

	case_GETOPT_HELP_CHAR;

	case_GETOPT_VERSION_CHAR (PROGRAM_NAME, AUTHORS);

	default:
	  usage (2);
	}
    }

  if (operating_mode == undefined_mode)
    FATAL_ERROR (_("you must specify a list of bytes, characters, or fields"));

  if (delim != '\0' && operating_mode != field_mode)
    FATAL_ERROR (_("a delimiter may be specified only when operating on fields"));

  if (suppress_non_delimited && operating_mode != field_mode)
    FATAL_ERROR (_("suppressing non-delimited lines makes sense\n\
\tonly when operating on fields"));

  if (!delim_specified)
    delim = '\t';

  if (output_delimiter_string == NULL)
    {
      static char dummy[2];
      dummy[0] = delim;
      dummy[1] = '\0';
      output_delimiter_string = dummy;
      output_delimiter_length = 1;
    }

  if (optind == argc)
    exit_status |= cut_file ("-");
  else
    for (; optind < argc; optind++)
      exit_status |= cut_file (argv[optind]);

  if (have_read_stdin && fclose (stdin) == EOF)
    {
      error (0, errno, "-");
      exit_status = 1;
    }

  exit (exit_status == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}
